YachtDice: implement new game

spinerak opened the pull request:
## What is this fixing or adding?
This is adding the game Yacht Dice. Yacht Dice is specifically made for AP. Yacht Dice is based on the public domain dice game called "Yacht". Yacht Dice can be played on browsers (also on mobile).

## How was this tested?
- This game has been tested often in unsupported sessions. There are no known issues.
- I have generated 1000 games with just one randomized Yacht Dice yaml (no issues).
- I have done a test with 1024 Yacht Dice yamls (no issues).
- I have done tests with Yacht Dice and other games (no issues).
- I ran the Archipelago unit tests. (no issues, TestAllGamesMultiworld was skipped: too slow for main)

Please let me know if I can assist, explain or help! @Spineraks on discord or via github.

NewSoupVi commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,223 @@
+        return mean_score * self.quantity
+
+
+def extract_progression(state, player, options):
+    #method to obtain a list of what items the player has.
+    #this includes categories, dice, rolls and score multiplier etc.
+    
+    if player == "state_is_a_list": #the state variable is just a list with the names of the items
```
So, as far as I understand it, every check in the game uses the same "custom state", which is what you calculate in `extract_progression`. I'll keep referring to return value of `extract_progression` as the "custom state" from here on out.

So, in any given location's access rule, the maximum achievable score is calculated and compared.
The maximum achievable score calculation is cached by the custom state. Very good.
However, and correct me if I'm wrong, but the custom state is not retained or cached in any way.

In my opinion, you should make the custom state (or maximum score, for that matter) persistent, and **override state.collect() and state.remove()**, and do the calculation of the custom state every time they get called with any (relevant) item - I.e. "on state change".

If you want to be really fancy, instead of recalcing the state every time `collect` or `remove` gets called, your overriden `state.collect()` and `state.remove()` could instead just set a custom_state_is_stale variable to `True`, and then check for that variable at the start of `extract_progression` to see if the custom state needs to be recalculated. 

I hope this makes sense.

NewSoupVi replied:
Btw, I would also recommend changing the "custom state" to a dataclass or NamedTuple, that would make it a bit easier to follow in my opinion. The dataclass could have an overriden str() method that then replaces your string representation conversion

NewSoupVi replied:
In fact, yeah, I think I'm convincing myself that this is the correct approach:

1. Make `maximum_achievable_score` a persistent state property, as it's the only thing that actually "matters" in the end
2. Make a second persistent state property called `achievable_score_is_stale`. If this is True, the maximum score needs to be recalculated.
3. Override collect & remove to set `achievable_score_is_stale` to True.

NewSoupVi replied:
Full disclosure, I've never done the collect/remove overriding myself, so I might be missing some nuance.

spinerak replied:
Thank you for the suggestions, I'll get working on it later today!
Yes you seem to understand how the code works!

I want to double check if I understand your suggestions.

First, here's some extra information.
I did a test run with Yacht Dice, SM64 and Celeste64, and here's the number of calls to the Yacht Dice logic:
- 29341 calls to `dice_simulation_strings`
- Out of all these calls, in 27386 of them, the state (of progression items, excl extra points) was exactly the same as the previous time the function was called
- In total, `dice_simulation_strings` was called with 296 unique states (of progression items, excl extra points)

I agree with your latest approach. One thing I'm planning to add to it, is that I only change `achievable_score_is_stale` to True, if a progressive item is added. 

I will also not change it to True when an Extra Point is added or removed, I will keep these outside of "maximum_achievable_score", and will make the logic be something like 
```
score_to_beat = 100 #example
extra_points_in_logic = state.count("1 Point", player)
extra_points_in_logic += state.count("10 Points", player) * 10
extra_points_in_logic += state.count("100 Points", player) * 100
if achievable_score_is_stale:
  update maximum achievable score
return state.has("maximum_achievable_score", player, score_to_beat - extra_points_in_logic)
```

Does this make sense to you?

NewSoupVi replied:
Yeah that seems fine :) state.count is quite fast.

If you ever have more denominations of point values though, it might be worth handling that through collect and remove as well, in a similar way to how DLCQuest works.

spinerak replied:
I've changed the extract_progression in Rules so that it doesn't do everything twice.
And as you know your idea works :)

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,404 @@
+        for item in self.precollected:
+            self.multiworld.push_precollected(self.create_item(item))
+
+    def create_items(self):
+        #convert strings to actual items
+        itempool_created = [item for item in map(lambda name: self.create_item(name), self.itempool)]
+
+        #and add them to the itempool
+        for item in itempool_created:
+            self.multiworld.itempool += [item]    
```
Ok I wasn't gonna comment on anything else yet before I'm ready to do a full review, but

*Please*. You're killing me here XD

```
def create_items(self):
    self.multiworld.itempool += [self.create_item(name) for name in self.itempool]
```

:D

spinerak replied:
Thanks, changed!

Exempt-Medic commented on worlds/yachtdice/Locations.py:
```diff
@@ -0,0 +1,67 @@
+    def __init__(self, player: int, name: str, score: int, address: typing.Optional[int], parent):
+        super().__init__(player, name, address, parent)
+        self.yacht_dice_score = score
+        self.event = not address
```
What is this doing? I don't see you using it anywhere and it is no longer a property of `Location`

spinerak replied:
Yup, thanks, I indeed don't use the event.
I've removed the `self.event = not address` line

Exempt-Medic commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,404 @@
+        
+    def pre_fill(self):
+        #in the pre_fill, make sure one dice and one roll is early, so that you will have 2 dice and 2 rolls soon
+        self.multiworld.early_items[self.player]["Dice"] = 1
```
Early items should be added in `generate_early`

spinerak replied:
OK, thanks, I've moved it to the end of `generate_early`

ScipioWright commented:
Don't forget to update codeowners and the archipelago readme

Exempt-Medic reviewed (changes requested):
A variety of small changes that should be made. Two bigger things though:

You may find it helpful to run `WebHost.py` and see how your game and its option names and descriptions display on the website. There are some unusual capitalization choices for displayed names that other games don't follow.

You may also find OptionGroups incredibly helpful for organizing the WebHost display. Grouping all six weight options together would be an excellent change in my opinion.

spinerak commented on worlds/yachtdice/Items.py:
```diff
@@ -0,0 +1,114 @@
+        "Category Five Distinct Dice",
+        "Category 4&5 Full House"
+    }
+}
```
Thank you for all the comments, I'll go through them :3

spinerak commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,270 @@
+    
+"""
+Test 1 2 3
+"""
```
Whoooops

spinerak commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,270 @@
+    Note: if you have extra points on "all_of_it", there will not be story chapters.
+    
+    all_of_it: fill all locations with story chapters
+    sure: if there is space left, put in 10 story chapters.
```
Yes! There is always at least 10 extra items.
`extra_locations_needed = max(10, math.ceil(already_items * extraPercentage))`

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,223 @@
+            state.count("Roll") 
+            + state.count("Roll Fragment") // options.number_of_roll_fragments_per_roll.value
+        )
+        roll = state.count("Roll") 
```
Nope I guess not... Good catch!

spinerak commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,270 @@
+    option_extreme = 4
+    default = 2
+        
+class scoreForLastCheck(Range):
```
I will think about this one. I agree with it being a bit annoying, but I like it being able to be every value, especially in random games.

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+            nonlocal weights
+            nonlocal extra_points_added
+            nonlocal multipliers_added
+            nonlocal items_added
```
WHAT IN THE WORLD IS NONLOCAL

spinerak replied:
chatgpt know what `nonlocal` is :3
I've changed it

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+            all_items = self.itempool + self.precollected
+            dice_fragments_in_pool = all_items.count("Dice") * frags_per_dice + all_items.count("Dice Fragment")
+            if dice_fragments_in_pool + 1 >= 9 * frags_per_dice:
+                weights[0] = 0  # don't allow >=9 dice
```
Hmmm

I wonder if this code would be clearer if `weights` was a `Dict[str, float]`
And you modified it like `weights["Dice"] = 0`

And then at the end, you do `which_item_to_add = self.random.choice(weights.keys(), weights=weights.values)`

And then after that, you can go ahead with `if which_item_to_add == "Dice"`

You feel me?

I think that would be a big improvement actually so I'm not gonna mark this Low Priority, I'd rather you do it or comment on why you don't want to

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                    weights[5] = 1
+
+            # Next, add the appropriate item. We'll slightly alter weights to avoid too many of the same item
+            which_item_to_add = self.multiworld.random.choices([0, 1, 2, 3, 4, 5], weights=weights)[0]
```
Please use self.random instead of self.multiworld.random

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                    return "Dice Fragment"
+            elif which_item_to_add == 1:
+                weights[1] /= 1 + frags_per_roll
+                if frags_per_roll == 1:
+                    return "Roll"
+                else:
+                    return "Roll Fragment"
```
I think this would look slightly nicer with a ternary?

`return "Roll" if frags_per_roll == 1 else "Roll Fragment"`

Personal preference tho

*Low Priority*

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                multipliers_added += 1
+                return "Step Score Multiplier"
+            elif which_item_to_add == 4:
+                cat_weights = [2, 2, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1]
```
Hmmm, I'm noticing a lot of your code is sort of like, inline lists that come out of nowhere for the reader

I think it'd be nicer if `cat_weights` was defined somewhere as a constant with the typing Dict[str, int] for example, like

```
# somewhere in, idk, YachtWeights.py? Or sth
CATEGORY_WEIGHTS = {
    "Category 1": 2,
    "Category 2": 2,
    ...
}
```

and then here, you did `self.random.choices(CATEGORY_WEIGHTS.keys(), weights=CATEGORY_WEIGHTS.values())`

Or maybe the weight could be a *property* of the Category class?
Then it'd be like

```
weights = [category.weight for category in possible_categories]
```

or something

Idk it's something to think about, I won't block the PR over stuff like this

*Low Priority*

spinerak replied:
This one is weird, because there are 16 categories, but every category can be the 'original' or the 'alt' one.
So the first category here will either be "Category Choice" OR "Double Threes and Fours". 
So the first idea wouldn't really work I think.
The second idea, with a property in the category class is interesting. But I feel like these numbers are not really properties of the categories, they are just more generation-parameters. So I'm going to use the _Low priority_ card here. I did add some comments to explain that these are the settings and why I've chosen them.

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+            elif which_item_to_add == 4:
+                cat_weights = [2, 2, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1]
+                weights[4] /= 1.1
+                return self.multiworld.random.choices(possible_categories, weights=cat_weights)[0]
```
Please use self.random instead of self.multiworld.random.

(I'm gonna stop mentioning that now, but please do it everywhere)

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                    extra_points_added += 100
+                    return "100 Points"
+                else:
+                    raise Exception("Unknown point value (Yacht Dice)")
```
Ok I mean I like that you're writing safe code but how in the heck would this ever happen

*Low Priority*

spinerak replied:
Haha maybe at some point it will be fun to add "2 Points" all the way up to "99 Points" as items. It felt a bit too random for now, but if I ever add it I'll definitely hit this exception.........

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                weights[4] /= 1.1
+                return self.multiworld.random.choices(possible_categories, weights=cat_weights)[0]
+            elif which_item_to_add == 5:
+                score_dist = self.options.points_size.value
+                probs = [1, 0, 0]
+                if score_dist == 1:
+                    probs = [0.9, 0.08, 0]
+                if score_dist == 2:
+                    probs = [0, 1, 0]
+                if score_dist == 3:
+                    probs = [0, 0.3, 0.7]
+                if score_dist == 4:
+                    probs = [0.3, 0.4, 0.3]
+                c = self.multiworld.random.choices([0, 1, 2], weights=probs)[0]
+                if c == 0:
+                    weights[5] /= 1.01
+                    extra_points_added += 1
+                    return "1 Point"
+                elif c == 1:
+                    weights[5] /= 1.1
+                    extra_points_added += 10
+                    return "10 Points"
+                elif c == 2:
+                    weights[5] /= 2
+                    extra_points_added += 100
+                    return "100 Points"
```
Again, I would find this so much nicer as:

```
score_dist = self.options.points_size.value

probs = { "1 Point": 1, "10 Points": 0, "100 Points": 0}
if score_dist == 1:
    probs = { "1 Point": 0.9, "10 Points": 0.08, "100 Points": 0}
... you get the idea ...

choice = self.random.choices(probs.keys(), weights=probs.values())[0]
if choice == "1 Point":
    weights["Extra Points"] /= 1.01
    extra_points_added += 1
elif choice == "10 Points":
    ...

return choice
```

Oh well like it's not the end of the world or anything

*Low priority*

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+                weights[4] /= 1.1
+                return self.multiworld.random.choices(possible_categories, weights=cat_weights)[0]
+            elif which_item_to_add == 5:
+                score_dist = self.options.points_size.value
```
Please do not compare against integers for `Choice` options.

There are other instances of this, please fix all of them.

Better alternatives are string comparisons (you can compare the option against the actual choice names, omitting the "option_" at the start, this is a special comparison we have programmed in):

```
score_dist = self.options.points_size  # no ".value"

if score_dist == "small":
    ...
```

or comparing against the constant itself:

```
score_dist = self.options.points_size  # no ".value"

if score_dist == PointsSize.option_small:
```

The latter is the neatest but requires importing PointsSize.
The former is also acceptable.

spinerak replied:
This was a spicy one 🌶️
I went from liking it to hating it, to liking it again.

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+        if score_in_logic > 1000:
+            removed_item = ""
+            while score_in_logic > 1000:
+                removed_item = self.itempool[-1]
+                self.itempool.pop()
```
You can just `removed_item = self.itempool.pop()` haha

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+            if loc_data.region == board.name
+        ]
+
+        # which index of all locations should have the Victory item.
+
```
(I think you removed some code here but forgot to remove the comment lol)

NewSoupVi commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+    def fill_slot_data(self):
+        # make slot data, which consists of yachtdice_data, options, and some other variables.
+        yacht_dice_data = self._get_yachtdice_data()
+        yacht_dice_options = self.options.as_dict(
```
(There's also probably a more elegant way to do that)

NewSoupVi replied:
I'm also ok with you just, like, removing the list you pass to `as_dict`

spinerak replied:
I only need 7 options for my game, so I guess I'll just put these in slot data.
It was nice for debugging but I guess going to PR it's nice to reduce slot data usage.

NewSoupVi reviewed (changes requested):
## Code Review:

Any comment I've made that ends with an *italicized* signature (e.g. "*Low Priority*"), you may click resolve on without doing any changes. Please consider the change first though.

This code is pretty good. It's a bit unpythonic in places, and a bit hard to read due to a bit of a case of abbreviation-itis and preferring numbers over strings / lists over dicts,

I had a lot to say but it's not because the code is bad, it was actually quite nice to review.

Some things do need to be changed though.

Did not review: Docs, the actual underlying math (just tried to think of ways to optimise what *is* being done)

NewSoupVi commented:
## Performance review

I can personally attest that effort has been made to improve the performance of this game. It's still not amazing, but I am personally out of ideas, other than small line-by-line optimisations. 1s for a single game is on the slow end, but it's not the slowest we have.

## Generation review

I am skipping this one this time because it has already been done by other people. Thanks!

spinerak commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+        frags_per_dice = self.options.number_of_dice_fragments_per_dice.value
+        frags_per_roll = self.options.number_of_roll_fragments_per_roll.value
+
+        # count number of plando items not from pool, we need extra locations for them
+        self.extra_plando_items = 0
+
+        for plando_setting in self.multiworld.plando_items[self.player]:
+            if plando_setting.get("from_pool", False) is False:
+                self.extra_plando_items += sum(value for value in plando_setting["items"].values())
```
Thanks for all the comments, all except this one are clear and I will start working on em!

The main culprit why I have this code is because someone added this to the yaml (with items plando in host.yaml on):
```
  plando_items:
    - items:
        Dice Fragment: 5
      from_pool: false
      world: true # I prefer those items to be placed in other worlds
      force: true
```
Just genning Yacht Dice with this yaml gives the error/warning
```
Unplaced items(5): [Encouragement (Player 1), Encouragement (Player 1), Encouragement (Player 1), Bonus Point (Player 1), Bonus Point (Player 1)] - Unfilled Locations(0): []
Per-Player counts: {'items': Counter({1: 85}), 'locations': Counter({1: 80})})
```
If I add 50 Dice Fragments instead, the gen fails because there is not enough locations for progression items.

I have asked in the discord for further feedback

spinerak replied:
removed :D

spinerak commented on worlds/yachtdice/Locations.py:
```diff
@@ -0,0 +1,73 @@
+        hiscore = curscore
+        scores += [curscore]
+
+    if goal_score != max_score:
+        # if the goal score is not in the list, find the closest one and make it the goal.
+        if goal_score not in scores:
+            closest_num = min(scores, key=lambda x: abs(x - 500))
+            scores[scores.index(closest_num)] = goal_score
```
Oof there is actually a mistake there, '500' should be goal_score. (it still works with 500 but not how I want :))
What this does, if goal_score is not one of the locations, it searches the closest one to goal_score and makes that location exactly the goal_score.
So if goal is 650 and there is 649 location, the 649 location will become 650 so that becomes the goal.

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+        self.state = state
+
+    def count(self, item: str, player: Optional[str] = None) -> int:
+        return self.state.count(item)
```
You may!
122.4s -> 121.0s for 100 games
We take those!

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+    number_of_fixed_mults = state.count("Fixed Score Multiplier", player)
+    number_of_step_mults = state.count("Step Score Multiplier", player)
+    categories = []
+    for category_name, category_value in category_mappings.items():
+        if state.count(category_name, player) >= 1:
+            categories += [Category(category_value, state.count(category_name, player))]
```
Thanks!
Yes please be as picky as you can be!

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+                combined_dist[val1 + val2] += prob1 * prob2
+        return dict(combined_dist)
+
+    # function to take the maximum of "times" i.i.d. dist1.
```
### Welcome to my territory! Glad you're asking :)
i.i.d. means independent(ly) and identical(ly) distributed.

So let's say you roll for Tiny Straight and you have a 30% chance to get it (20 points).
Now you could write this as a "random variable", say $X$, which is 0 with probability 0.7 and 20 with probability 0.3. $X$ is a (probability) distribution.

Now what if you would get two shots at the Tiny Straight? Mathematically, you could define two **i.i.d. random variables** of $X$, let's call them $X_1$ and $X_2$. 
If we get two shots, we are interested in $\max(X_1, X_2)$, which is also a random variable, in this case:
$\max(X_1, X_2) = 0$ with probability $0.7 \cdot 0.7$
$\max(X_1, X_2) = 20$ with probability $1 - 0.7 \cdot 0.7$
Although you would probably define $Z = \max(X_1, X_2)$ and say $\mathbb{P}[Z = 20] = 1- {0.7}^2$ and 0 otherwise.
 
Later on in the logic we also sum random variables. For example $X$ and $Y$. If $X$ is $i$ with probability $p_i$ and $Y$ is $j$ with probability $q_j$, $$\mathbb{P}[X+Y=k] = \sum_i  p_i \cdot q_{k-i}$$. This sums all possibilities that $X$ could be ($i$), and then puts $Y$ to be $k-i$, so that X+Y equals $i+k-i = k$.

I'm abusing notation a little bit but I'm sure the maths Gods still appreciate my effort.

Now what was I doing........

NewSoupVi replied:
HELP

NewSoupVi replied:
WHERE IS THE RESOLVE BUTTON AAAAAAAAAAAAAAAAA

spinerak commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+        categorylist = [1] * num_ones + [0] * (16 - num_ones)
+
+        # Shuffle the list to randomize the order
+        self.multiworld.random.shuffle(categorylist)
```
Cool I changed it everywhere 👍

spinerak commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+
+        already_items = len(self.itempool) + self.extra_plando_items
+
+        # Yacht Dice needs extra filler items so it doesn't get stuck in generation.
+        # For now, we calculate the number of extra items we'll need later.
+        if self.options.minimize_extra_items.value:
+            extra_percentage = max(0.1, 0.8 - self.multiworld.players / 10)
+        else:
+            extra_percentage = 0.7
+        extra_locations_needed = max(10, math.ceil(already_items * extra_percentage))
```
I made this an option because reducing filler items might break if you join Yacht Dice with 6 Cliques. But any game with more than 2 locations should be fine.

spinerak commented:
The comments are super helpful, I'll continue later this week.

I have a couple of questions due to my selective perfectionism if you don't mind @NewSoupVi 
- Is it OK that there is a "Victory" **item**? It works really well (in generating, in game, in log), but I see now in the webhost that you can pick it as "starting inventory". If you do and when you open the game, it immediately 'wins' and 'releases' all items. To remedy it, I could make "Victory" an event or I could maybe force-remove Victory from starting inventory. Or I can leave it as is, what do you think?
- On the AP website, my locations are ordered 1 Score, 10 Score, 100 Score, 101 Score, ..., 199 Score, 2 Score, 20 Score etc. Do you think I should name my locations 001 Score, 002 Score etc, so that it is ordered? I'd probably make the last location 999 then.

Exempt-Medic commented:
> The comments are super helpful, I'll continue later this week.
> 
> I have a couple of questions due to my selective perfectionism if you don't mind @NewSoupVi
> 
> * Is it OK that there is a "Victory" **item**? It works really well (in generating, in game, in log), but I see now in the webhost that you can pick it as "starting inventory". If you do and when you open the game, it immediately 'wins' and 'releases' all items. To remedy it, I could make "Victory" an event or I could maybe force-remove Victory from starting inventory. Or I can leave it as is, what do you think?
> * On the AP website, my locations are ordered 1 Score, 10 Score, 100 Score, 101 Score, ..., 199 Score, 2 Score, 20 Score etc. Do you think I should name my locations 001 Score, 002 Score etc, so that it is ordered? I'd probably make the last location 999 then.

Typically games make the Victory item an event, this allows it to show up in the spoiler log and prevents it from being put into things like start inventory or being obtained with `!get item`

For the item name sorting, I personally would say that's a problem with the way the website sorts them and not with your game. You could choose to work around it, but I think the WebHost should make a correction

spinerak commented on worlds/yachtdice/Locations.py:
```diff
@@ -0,0 +1,73 @@
+    # to avoid repeating scores, hiscore keeps tracks of the highest score location
+    # and the next score will always be at least hiscore + 1
+    # note that curscore is at most max_score-1
+    hiscore = 0
+    for i in range(num_locs - 1):
+        perc = i / num_locs
+        curscore = int(1 + (perc**scaling) * (max_score - 2))
+        if curscore <= hiscore:
+            curscore = hiscore + 1
+        hiscore = curscore
+        scores += [curscore]
```
The main thing was I needed to have exactly `number_of_locations` locations, so I couldn't quite make that work if I had to remove duplicates.
This btw leads to Yacht Dice having quite a few locations at the start, but this also seems necessary to assure Yacht Dice solo can be completed

spinerak commented on worlds/yachtdice/Locations.py:
```diff
@@ -0,0 +1,73 @@
+
+
+all_locations = {}
+starting_index = 16871244500  # 500 more than the starting index for items
```
Oh, I think I took inspiration from the manual apworld which does something like this iirc.
I added a comment that it's not really necessary to do this.
If you think it's wise (for performance maybe) to change this to starting_index 0, let me know :)

NewSoupVi replied:
Nah, it's fine. You'd have to change the client too, so at this point, it's alright. And we're now in per-game-ID land anyway.
Maybe when you add some new items, you can lower the starting index to #000, and then add 500 to the existing items, and add the new items at the beginning... or sth. Lol
But yeah this is not really important

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+        cumulative_prob = 0
+        prev_val = None
+
+        for val in sorted_values:
+            prev_val = val
+            cumulative_prob += dist[val]
+            if cumulative_prob >= percentile:
+                return prev_val  # Return the value before reaching the desired percentile
```
Good catch, yeah this is a 'mistake', I wanted it to do the first, but it is indeed doing the second.
With all the testing and calibration, I won't change it to the first, instead I will simplify it to the second.

Two lines later, I also put 
```
return sorted_values[-1]
```
which would only occur if percentile would somehow be larger than 1.

spinerak commented on worlds/yachtdice/Locations.py:
```diff
@@ -0,0 +1,73 @@
+lookup_id_to_name: typing.Dict[int, str] = {data.id: item_name for item_name, data in all_locations.items() if data.id}
+
+# we need to run this function to initialize all scores from 1 to 1000, even though not all are used
+all_locations = all_locations_fun(1000)
```
I don't use lookup_id_to_name at all, so I just removed that variable.
Unless it is one of the AP necessary functions, but afaik these are only item_name_to_id and location_name_to_id.

NewSoupVi replied:
Yeah you can remove this if it's unused

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+def dice_simulation_strings(categories, num_dice, num_rolls, fixed_mult, step_mult, diff):
+    tup = tuple(
+        [
+            tuple(sorted([c.name + str(c.quantity) for c in categories])),
```
Yeah sure I'll, try, just know that I will have less time to tell you all about the wondrous world of probability distributions! I was just about to introduce the probability generating function.......

Anyways I have tried what you mentioned but it is x2 slower.
Can you check if I did what you meant? Here is the commit (I reverted it for now):
https://github.com/ArchipelagoMW/Archipelago/pull/3482/commits/34f2c1aed8c8813b2d9c58896650b82a810d3578

NewSoupVi replied:
Awwww, okay. Yeah, if it's slower, then it's not worth it obv :)

NewSoupVi replied:
Oh, the reason I was saying it is because I thought maybe then you could just completely get rid of

```
tup = (
     categories,
     num_dice,
     num_rolls,
     fixed_mult,
     ...
)
```

because now that's just the output of `extract_progression` directly. Like essentially, maybe at this point you could just

```
tup = extract_progression(...)
```

But I don't imagine this will speed things up by a bunch suddenly, so

spinerak replied:
A bit further we sort the tuple/list by meanscore. It seems that sorting a tuple is slower than sorting a list, maybe?

NewSoupVi replied:
Oh, maybe. Well, I tried! Feel free to go without this then :)

NewSoupVi commented:
> The comments are super helpful, I'll continue later this week.
> 
> I have a couple of questions due to my selective perfectionism if you don't mind @NewSoupVi
> 
> * Is it OK that there is a "Victory" **item**? It works really well (in generating, in game, in log), but I see now in the webhost that you can pick it as "starting inventory". If you do and when you open the game, it immediately 'wins' and 'releases' all items. To remedy it, I could make "Victory" an event or I could maybe force-remove Victory from starting inventory. Or I can leave it as is, what do you think?
> * On the AP website, my locations are ordered 1 Score, 10 Score, 100 Score, 101 Score, ..., 199 Score, 2 Score, 20 Score etc. Do you think I should name my locations 001 Score, 002 Score etc, so that it is ordered? I'd probably make the last location 999 then.

As Medic said, having a Victory item is not necessary. Having a Victory **event** item (address = None) is nice because then Victory displays in the spoiler log.

Exempt-Medic commented:
> To do (note for myself): minecraft Dig gave an error with Yacht Dice: WARNING: Error proccessing incoming packet: java.lang.NumberFormatException: For input string: "16871244127" Probably because my ids are larger than what ints could be. Not sure if Yacht Dice problem, but might as well change my ids now. I've let the dev of minecraft Dig know as well, Zork grand inquisitor also has ids that do not fit in ints.

https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/world%20api.md#locations
> World-specific IDs must be in the range 1 to 2<sup>53</sup>-1.

This is a Minecraft Dig problem

spinerak commented on worlds/yachtdice/Rules.py:
```diff
@@ -0,0 +1,253 @@
+# We will store the results of this function as it is called often for the same parameters.
+
+
+yachtdice_cache = {}
```
That's a good question.
Every world stores around 100~200 results in the yachtdice_cache. The max I've seen is 400. There seems to be _almost no overlap_ between worlds.
Would it be smart to have a separate cache per world? Is that safer in terms of server explosions?

spinerak replied:
I did some experiments with 236 Yacht Dice worlds.
- I tried out what a separate cache per world would do. It increased the generation time from 234 to ~240 seconds. I think that's fine.
- I tried what would happen when you cap the size of the cache to 400 entries _per player_. This did not affect the speed at all. 

I have no idea what's smart or what fits the AP project...

spinerak replied:
Actually, I've tested with 1024 Yacht Dice worlds. And splitting the cache per player is actually faster there (probably saves a lot of time searching the cache).

spinerak commented on worlds/yachtdice/__init__.py:
```diff
@@ -0,0 +1,459 @@
+
+        # Add the victory item to the correct location.
+        # The website declares that the game is complete when the victory item is obtained.
+        board.locations[goal_index].place_locked_item(self.create_item("Victory"))
```
In that case, I'd rather not return the goal score at all.
```
victory_location_name = f"{self.goal_score} score"
self.get_location(victory_location_name).place_locked_item(self.create_item("Victory"))
```
I think this 'return' was quite old and I don't need it anymore :)

spinerak commented:
I went through all of your comments @NewSoupVi , thank you for all of them!
I changed almost everything except one or two low priority comments, I did comment on them.

Here's the current status and open points, from my side, with the <ins>most important parts</ins> underlined:
- I've changed the weights to speed up generation by ~40%. My mathematics heart is broken a little bit, but the differences are really small (the game's logic is now sliiiightly stricter, which makes the game sliiiightly easier, but only by at most 10 points out of 1000 I think). So <ins>I'm happy with the current state</ins>, without numpy. Numpy would still speed up my game of course.
- Yacht dice cache: you asked how big the cache would be. I answered here https://github.com/ArchipelagoMW/Archipelago/pull/3482#discussion_r1639033215 and the message after. For now <ins>I separated the cache per player and limit it to 400</ins>. But if size of cache doesn't matter that much, <ins>one cache with one limit would be the fastest</ins>. What do you think?
- The Victory item. I've taken a quick look into the Javascript client and I can't get it to work (I know the status is a little broken, it won't say 'disconnected' for example). From the client point of view, I like checking for the Victory item in inventory, more than 1) asking the status on game opening and 2) reading JSON events and checking if there is a game win for the current game. As such, <ins>I would prefer keeping the Victory item</ins>. Let me know if you disagree and I can take another look. My main concern is ~~memers~~ players will put the Victory item in start_inventory, but they could also put 1000 points in the start_inventory which would have the same effect.
- I looked into the webhost, and there will be sliders for options, ranging from values from 500 to 1000. It will be awkward to select a specific value (you can use arrow keys, but it's unintuitive). It was suggested to only have a select few values to choose from. I would like to keep all values in the option though, this is really fun when you put it to random. So I prefer to <ins>keep the current settings</ins>. Again, let me know if you disagree.

ReverM commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,331 @@
+
+class ChanceOfPoints(Range):
+    """
+    Getting points gives you... points. You can get 1 point, 10 points, and even 100 points.
```
Unclear what this option does. Could the description be rephrased or clarified?

spinerak replied:
Thank you for the feedback!
I've resolved the other comments (changed automatically now defines what it changes, and I've rephrased the questions in the yaml, I agree on those).

I've changed the description of this option to:
    Are you tired of rolling dice countless times and tallying up points one by one, all by yourself? 
    Worry not, as this option will simply add some points items to the item pool! 
    And getting one of these points items gives you... points!
    Imagine how nice it would be to find tons of them. Or even better, having others find them FOR you!
I think it's funny _and_ clear this way. Do you agree (with the fact that it's clear 😄)?

ReverM commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,331 @@
+class ScoreForGoal(Range):
+    """
+    This option determines what score you need to reach to finish the game.
+    It cannot be higher than the score for the last check (if it is, it is changed automatically).
```
Might be okay but "it is changed automatically", which of the two option mentionned is changed?

ReverM commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,331 @@
+class AddExtraPoints(Choice):
+    """
+    Yacht Dice typically has space for extra items.
+    If there is space, would you like bonus points shuffled in the item pool?
```
```
    If there is space, would you like bonus points shuffled in the item pool?
```
Having a question in the description feels odd. I would potentially rephrase it like "you can add bonus points to the item pool" instead of a question

ReverM commented on worlds/yachtdice/Options.py:
```diff
@@ -0,0 +1,331 @@
+    Yacht Dice allows players to roll IRL dice.
+    By sending "manual" in the chat, an input field appears where you can type your dice rolls.
+    Of course, we cannot check anymore if the player is playing fair.
+    Do you want to allow manual input of rolls?
```
Same as above with the other questions

ReverM reviewed (commented):
I have only looked at options description and docs. 
Some slightly weird formulations in the option description
Docs LGTM
